File: subjects/PDOO/Exámenes de Teoría/Examen 02/Examen.java
import java.util.Random;

public class Examen{
    private float atributo1;
    private float atributo2;
    private static final int MAX_SUMA_ATRIB=10;

    public Examen(float param1, float param2){
        atributo1=param1;
        atributo2=param2;

        while(!(cumpleRestricciones(param1) && cumpleRestricciones(param2))){
            // Código que modifica los atributos
        }
    }

    private boolean cumpleRestricciones(float a){
        boolean result=true;
        // Código que modifica el valor de result
        return result;
    }


    // Método rand mencionado. No se aportaba en el examen porque no era necesario.
    private static float rand(int max){
        Random r= new Random();
        return r.nextFloat()*max;
    }


    // Ejercicio 1
    /*
    void metodoExamen(Examen e){
        e.cumpleRestricciones(e.atributo1);
        Examen e2= new Examen(1,2);
        e2.cumpleRestricciones(e2.atributo1);
    }
     */



    // Ejercicio 3
    /*
    Examen(){
        this(rand(MAX_SUMA_ATRIB), rand(MAX_SUMA_ATRIB));
    }

    // Método main
    public static void main(String[] args){
        Examen e = new Examen(3, 4);    // Constructor actual
        Examen e2= new Examen();        // Constructor sin parámetros
    }
     */
    
}


// Ejercicio 5
// Para comprobar que da error, se ha de comentar el constructor sin parámetros del Ejercicio 3
// El ejercicio 8 no pertenece a este ejercicio, por lo que ha de comentarse.
/*
class ExamenDerivada extends Examen{

    // Ejercicio 8
    @Override
    public boolean cumpleRestricciones(float a){
        boolean resultado=false;
        // Cuerpo del método
        return resultado;
    }
    
}
 */

File: subjects/PDOO/Exámenes de Teoría/Examen 02/Examen.rb
class Examen
  @atributo1 = 1
  @atributo2 = 2
  @atributo3 = 3
  @MAX_SUMA_ATRIB = 10

  def initialize(param1, param2)
    @atributo1 = param1
    @atributo2 = param2

    while(!(cumple_restricciones(param1) && cumple_restricciones(param2)))
      # Código que modifica los atributos
    end
  end

  private

  def cumple_restricciones(a)
    result = true
    # Código que modifica el valor de result
    result
  end



  # Ejericio 2
  def metodoExamen(e)
    e.cumple_restricciones(7)
    e2= Examen.new(1,2)
    e2.cumple_restricciones(7)
  end


  # Ejercicio 4
  def self.newCons
    param1 = rand(@MAX_SUMA_ATRIB)
    param2 = rand(@MAX_SUMA_ATRIB)
    new(param1, param2)
  end


end


# Para probar el código
e = Examen.new(1,2)

# e.metodoExamen(e)     # Ejercicio 2

e2= Examen.newCons    # Ejercicio 4


# Ejercicio 6
class ExamenDerivada < Examen

end
ed = ExamenDerivada.new(1,2)
